File: JavaScript/15-Greedy/Algorithms/3-Kruskal’sAlgorithm(MST)/Kruskal’sAlgorithm(MST).js
// Kruskal's Algorithm for Minimum Spanning Tree (MST)

/*
Kruskal's algorithm is a greedy algorithm used to find a minimum spanning tree (MST) for a weighted, undirected graph. It works by repeatedly adding the cheapest edge to the MST that does not create a cycle.

Here's a breakdown of the algorithm:

1.  Sort all the edges of the graph in non-decreasing order of their weight.
2.  Create an empty MST.
3.  Pick the smallest edge. Check if including this edge in the MST creates a cycle.
    * If a cycle is not formed, include this edge in the MST.
    * Else, discard this edge.
4.  Repeat step 3 until there are (V-1) edges in the MST, where V is the number of vertices in the graph.

Data Structures:

* Graph Representation:  Adjacency List or Edge List.  For Kruskal's, an Edge List is generally preferred due to the sorting step.  An edge list is an array of tuples (u, v, weight) representing the edges.
* Disjoint Set Union (DSU):  Essential for cycle detection.  We'll implement this using:
    * Parent Array (for basic DSU)
    * Rank and Path Compression (for optimized DSU)

Time Complexity: O(E log E) or O(E log V), where E is the number of edges and V is the number of vertices.
    * Sorting edges takes O(E log E).
    * DSU operations (find and union) take nearly constant time with rank and path compression.
Space Complexity: O(V + E)
    * O(V) for the parent array in DSU.
    * O(E) for storing the edges.

Variations in Implementation (Demonstrated Below):

We'll show 5 different ways to implement Kruskal's, primarily focusing on variations in:
1.  Graph Representation: Edge List
2.  Disjoint Set Union (DSU) Implementation:
    * Basic Parent Array
    * Rank Optimization
    * Path Compression Optimization
    * Rank and Path Compression Optimization
    * ES6 Class based implementation of DSU
*/

// Example Graph (for all implementations)
const graph = [
    [0, 1, 10], // Edge between node 0 and 1 with weight 10
    [0, 2, 6],
    [0, 3, 5],
    [1, 3, 15],
    [2, 3, 4],
];
const numVertices = 4; // Number of vertices in the graph

// ------------------------------------------------------------------------------------------------
// 1. Kruskal's Algorithm with Basic Disjoint Set Union (Parent Array)
// ------------------------------------------------------------------------------------------------
function kruskalBasicDSU(edges, numVertices) {
    let mstWeight = 0;
    const parent = new Array(numVertices);

    // Initialize parent array: Each vertex is its own parent initially
    for (let i = 0; i < numVertices; i++) {
        parent[i] = i;
    }

    // Function to find the parent of a vertex (without path compression)
    function find(i) {
        if (parent[i] === i) {
            return i;
        }
        return find(parent[i]); // Recursive call
    }

    // Function to unite two sets (without rank)
    function union(i, j) {
        const a = find(i);
        const b = find(j);
        if (a !== b) { // only union if they are not already in the same set.
            parent[a] = b;
        }
    }

    // 1. Sort edges by weight in ascending order
    edges.sort((a, b) => a[2] - b[2]);

    const mstEdges = []; // Array to store edges of the MST

    // 2. Iterate through sorted edges
    for (const edge of edges) {
        const u = edge[0];
        const v = edge[1];
        const weight = edge[2];

        // 3. Find the parent of the vertices
        const parentU = find(u);
        const parentV = find(v);

        // 4. If including the edge doesn't form a cycle, include it in the MST
        if (parentU !== parentV) {
            mstWeight += weight;
            union(parentU, parentV); // Merge the sets
            mstEdges.push(edge);
        }
    }
    return { weight: mstWeight, edges: mstEdges };
}

// ------------------------------------------------------------------------------------------------
// 2. Kruskal's Algorithm with Rank Optimization in DSU
// ------------------------------------------------------------------------------------------------
function kruskalRankDSU(edges, numVertices) {
    let mstWeight = 0;
    const parent = new Array(numVertices);
    const rank = new Array(numVertices).fill(0); // Initialize rank array

    // Initialize parent and rank
    for (let i = 0; i < numVertices; i++) {
        parent[i] = i;
        rank[i] = 0;
    }

    function find(i) {
        if (parent[i] === i) {
            return i;
        }
        return find(parent[i]);
    }

    function union(i, j) {
        const a = find(i);
        const b = find(j);
        if (a !== b) {
            if (rank[a] < rank[b]) {  // Attach smaller rank tree under root of high rank tree
                parent[a] = b;
            } else if (rank[a] > rank[b]) {
                parent[b] = a;
            } else { // If ranks are same, then make one as root and increment its rank by one
                parent[b] = a;
                rank[a]++;
            }
        }
    }

    edges.sort((a, b) => a[2] - b[2]);
    const mstEdges = [];

    for (const edge of edges) {
        const u = edge[0];
        const v = edge[1];
        const weight = edge[2];

        const parentU = find(u);
        const parentV = find(v);

        if (parentU !== parentV) {
            mstWeight += weight;
            union(parentU, parentV);
            mstEdges.push(edge);
        }
    }
    return { weight: mstWeight, edges: mstEdges };
}

// ------------------------------------------------------------------------------------------------
// 3. Kruskal's Algorithm with Path Compression Optimization in DSU
// ------------------------------------------------------------------------------------------------
function kruskalPathCompressionDSU(edges, numVertices) {
    let mstWeight = 0;
    const parent = new Array(numVertices);

    // Initialize parent
    for (let i = 0; i < numVertices; i++) {
        parent[i] = i;
    }

    function find(i) {
        if (parent[i] === i) {
            return i;
        }
        // Path compression:  Make the parent of i the root of the tree
        parent[i] = find(parent[i]);
        return parent[i];
    }

    function union(i, j) {
        const a = find(i);
        const b = find(j);
        if (a !== b) {
            parent[a] = b;
        }
    }

    edges.sort((a, b) => a[2] - b[2]);
    const mstEdges = [];

    for (const edge of edges) {
        const u = edge[0];
        const v = edge[1];
        const weight = edge[2];

        const parentU = find(u);
        const parentV = find(v);

        if (parentU !== parentV) {
            mstWeight += weight;
            union(parentU, parentV);
            mstEdges.push(edge);
        }
    }
    return { weight: mstWeight, edges: mstEdges };
}

// ------------------------------------------------------------------------------------------------
// 4. Kruskal's Algorithm with Rank and Path Compression Optimization in DSU
// ------------------------------------------------------------------------------------------------
function kruskalOptimizedDSU(edges, numVertices) {
    let mstWeight = 0;
    const parent = new Array(numVertices);
    const rank = new Array(numVertices).fill(0);

    // Initialize parent and rank
    for (let i = 0; i < numVertices; i++) {
        parent[i] = i;
        rank[i] = 0;
    }

    function find(i) {
        if (parent[i] === i) {
            return i;
        }
        // Path compression:
        parent[i] = find(parent[i]);
        return parent[i];
    }

    function union(i, j) {
        const a = find(i);
        const b = find(j);
        if (a !== b) {
            if (rank[a] < rank[b]) {
                parent[a] = b;
            } else if (rank[a] > rank[b]) {
                parent[b] = a;
            } else {
                parent[b] = a;
                rank[a]++;
            }
        }
    }

    edges.sort((a, b) => a[2] - b[2]);
    const mstEdges = [];

    for (const edge of edges) {
        const u = edge[0];
        const v = edge[1];
        const weight = edge[2];

        const parentU = find(u);
        const parentV = find(v);

        if (parentU !== parentV) {
            mstWeight += weight;
            union(parentU, parentV);
            mstEdges.push(edge);
        }
    }
    return { weight: mstWeight, edges: mstEdges };
}

// ------------------------------------------------------------------------------------------------
// 5. Kruskal's Algorithm with ES6 Class for DSU
// ------------------------------------------------------------------------------------------------

class DisjointSet {
    constructor(numVertices) {
        this.parent = new Array(numVertices);
        this.rank = new Array(numVertices).fill(0);
        for (let i = 0; i < numVertices; i++) {
            this.parent[i] = i;
        }
    }

    find(i) {
        if (this.parent[i] === i) {
            return i;
        }
        this.parent[i] = this.find(this.parent[i]); // Path compression
        return this.parent[i];
    }

    union(i, j) {
        const a = this.find(i);
        const b = this.find(j);
        if (a !== b) {
            if (this.rank[a] < this.rank[b]) {
                this.parent[a] = b;
            } else if (this.rank[a] > this.rank[b]) {
                this.parent[b] = a;
            } else {
                this.parent[b] = a;
                this.rank[a]++;
            }
        }
    }
}

function kruskalClassDSU(edges, numVertices) {
    let mstWeight = 0;
    const dsu = new DisjointSet(numVertices);
    edges.sort((a, b) => a[2] - b[2]);
    const mstEdges = [];

    for (const edge of edges) {
        const u = edge[0];
        const v = edge[1];
        const weight = edge[2];

        const parentU = dsu.find(u);
        const parentV = dsu.find(v);

        if (parentU !== parentV) {
            mstWeight += weight;
            dsu.union(parentU, parentV);
            mstEdges.push(edge);
        }
    }
    return { weight: mstWeight, edges: mstEdges };
}

// ------------------------------------------------------------------------------------------------
// Output the results using console.log
// ------------------------------------------------------------------------------------------------

console.log("Graph:", graph);
console.log("Number of Vertices:", numVertices);

console.log("\n1. Kruskal's Algorithm with Basic DSU:");
const mstBasic = kruskalBasicDSU(graph, numVertices);
console.log("MST Weight:", mstBasic.weight);
console.log("MST Edges:", mstBasic.edges);

console.log("\n2. Kruskal's Algorithm with Rank Optimization:");
const mstRank = kruskalRankDSU(graph, numVertices);
console.log("MST Weight:", mstRank.weight);
console.log("MST Edges:", mstRank.edges);

console.log("\n3. Kruskal's Algorithm with Path Compression:");
const mstPathCompression = kruskalPathCompressionDSU(graph, numVertices);
console.log("MST Weight:", mstPathCompression.weight);
console.log("MST Edges:", mstPathCompression.edges);

console.log("\n4. Kruskal's Algorithm with Rank and Path Compression:");
const mstOptimized = kruskalOptimizedDSU(graph, numVertices);
console.log("MST Weight:", mstOptimized.weight);
console.log("MST Edges:", mstOptimized.edges);

console.log("\n5. Kruskal's Algorithm with Class-based DSU:");
const mstClass = kruskalClassDSU(graph, numVertices);
console.log("MST Weight:", mstClass.weight);
console.log("MST Edges:", mstClass.edges);

File: C++/15-Greedy/Algorithms/3-Kruskal’sAlgorithm(MST)/Kruskal’sAlgorithm(MST).cpp
// Kruskal's Algorithm - Minimum Spanning Tree (MST) using 5 approaches in C++
// Approaches differ by union-find optimizations and graph representations

#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>

using namespace std;

// Edge structure to store (weight, source, destination)
struct Edge {
    int u, v, weight;
    bool operator<(const Edge& other) const {
        return weight < other.weight;
    }
};

// Disjoint Set Union (DSU) with path compression and union by rank
class DSU {
private:
    vector<int> parent, rank;

public:
    DSU(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; ++i) parent[i] = i;
    }

    int find(int x) {
        if (x != parent[x])
            parent[x] = find(parent[x]); // Path compression
        return parent[x];
    }

    void unite(int x, int y) {
        int xr = find(x);
        int yr = find(y);
        if (xr == yr) return;

        if (rank[xr] < rank[yr])
            parent[xr] = yr;
        else if (rank[xr] > rank[yr])
            parent[yr] = xr;
        else {
            parent[yr] = xr;
            rank[xr]++;
        }
    }
};

// Approach 1: Kruskal's using edge list and DSU with full optimizations
int kruskal1(int n, vector<Edge>& edges) {
    sort(edges.begin(), edges.end());
    DSU dsu(n);
    int mst_weight = 0;

    for (Edge& e : edges) {
        if (dsu.find(e.u) != dsu.find(e.v)) {
            dsu.unite(e.u, e.v);
            mst_weight += e.weight;
        }
    }
    return mst_weight;
}

// Approach 2: Kruskal's using edge list and DSU with only path compression
class DSUPathCompressionOnly {
private:
    vector<int> parent;

public:
    DSUPathCompressionOnly(int n) {
        parent.resize(n);
        for (int i = 0; i < n; ++i) parent[i] = i;
    }

    int find(int x) {
        if (x != parent[x])
            parent[x] = find(parent[x]);
        return parent[x];
    }

    void unite(int x, int y) {
        parent[find(x)] = find(y);
    }
};

int kruskal2(int n, vector<Edge>& edges) {
    sort(edges.begin(), edges.end());
    DSUPathCompressionOnly dsu(n);
    int mst_weight = 0;

    for (Edge& e : edges) {
        if (dsu.find(e.u) != dsu.find(e.v)) {
            dsu.unite(e.u, e.v);
            mst_weight += e.weight;
        }
    }
    return mst_weight;
}

// Approach 3: Kruskal's with DSU having no optimizations (naive find/unite)
class DSUNaive {
private:
    vector<int> parent;

public:
    DSUNaive(int n) {
        parent.resize(n);
        for (int i = 0; i < n; ++i) parent[i] = i;
    }

    int find(int x) {
        while (x != parent[x]) x = parent[x];
        return x;
    }

    void unite(int x, int y) {
        parent[find(x)] = find(y);
    }
};

int kruskal3(int n, vector<Edge>& edges) {
    sort(edges.begin(), edges.end());
    DSUNaive dsu(n);
    int mst_weight = 0;

    for (Edge& e : edges) {
        if (dsu.find(e.u) != dsu.find(e.v)) {
            dsu.unite(e.u, e.v);
            mst_weight += e.weight;
        }
    }
    return mst_weight;
}

// Approach 4: Kruskal's with adjacency matrix converted to edge list
int kruskal4(int n, vector<vector<int>>& adjMatrix) {
    vector<Edge> edges;
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (adjMatrix[i][j] != 0)
                edges.push_back({i, j, adjMatrix[i][j]});
        }
    }
    return kruskal1(n, edges);
}

// Approach 5: Kruskal's using tuple representation for edges
int kruskal5(int n, vector<tuple<int, int, int>>& edgeTuples) {
    vector<Edge> edges;
    for (auto& [w, u, v] : edgeTuples) {
        edges.push_back({u, v, w});
    }
    return kruskal1(n, edges);
}

// Example usage:
int main() {
    int n = 4; // Number of vertices
    vector<Edge> edges = {
        {0, 1, 10}, {0, 2, 6}, {0, 3, 5}, {1, 3, 15}, {2, 3, 4}
    };

    cout << "Kruskal MST Approach 1: " << kruskal1(n, edges) << endl;
    cout << "Kruskal MST Approach 2: " << kruskal2(n, edges) << endl;
    cout << "Kruskal MST Approach 3: " << kruskal3(n, edges) << endl;

    vector<vector<int>> adjMatrix = {
        {0, 10, 6, 5},
        {10, 0, 0, 15},
        {6, 0, 0, 4},
        {5, 15, 4, 0}
    };
    cout << "Kruskal MST Approach 4 (adjacency matrix): " << kruskal4(n, adjMatrix) << endl;

    vector<tuple<int, int, int>> edgeTuples = {
        {10, 0, 1}, {6, 0, 2}, {5, 0, 3}, {15, 1, 3}, {4, 2, 3}
    };
    cout << "Kruskal MST Approach 5 (tuple-based): " << kruskal5(n, edgeTuples) << endl;

    return 0;
}
